File: source/go/counting_sort.go
package main

import (
	"fmt"
	"strings"
)

func countingSort(arr []int) {
	n := len(arr)
	if n == 0 {
		return
	}
	maxVal := arr[0]
	for i := 1; i < n; i++ {
		if arr[i] > maxVal {
			maxVal = arr[i]
		}
	}
	count := make([]int, maxVal+1)
	for i := 0; i < n; i++ {
		count[arr[i]]++
	}
	idx := 0
	for i := 0; i <= maxVal; i++ {
		for count[i] > 0 {
			arr[idx] = i
			idx++
			count[i]--
		}
	}
}

func main() {
	arr := []int{64, 34, 25, 12, 22, 11, 90, 1, 45, 78}
	countingSort(arr)
	parts := make([]string, len(arr))
	for i, v := range arr {
		parts[i] = fmt.Sprintf("%d", v)
	}
	fmt.Println(strings.Join(parts, " "))
}

File: source/python/counting_sort.py
def counting_sort(arr):
    n = len(arr)
    if n == 0:
        return arr
    max_val = max(arr)
    count = [0] * (max_val + 1)
    for x in arr:
        count[x] += 1
    idx = 0
    for i in range(max_val + 1):
        while count[i] > 0:
            arr[idx] = i
            idx += 1
            count[i] -= 1
    return arr

arr = [64, 34, 25, 12, 22, 11, 90, 1, 45, 78]
counting_sort(arr)
print(" ".join(str(x) for x in arr))
